sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -362,17 +374,13 @@ final class Decimal extends Ordered[Decimal] with Serializable {
   def abs: Decimal = if (this.compare(Decimal.ZERO) < 0) this.unary_- else this
 
   def floor: Decimal = if (scale == 0) this else {
-    val value = this.clone()
-    value.changePrecision(
-      DecimalType.bounded(precision - scale + 1, 0).precision, 0, ROUND_FLOOR)
-    value
+    toPrecision(DecimalType.bounded(precision - scale + 1, 0).precision, 0, ROUND_FLOOR)
```

Comment:
Theoretically, it should be `Some`. On the other hand if something goes wrong when setting new precision in `floor` or `ceil`, I would raise an internal error:

    def floor: Decimal = if (scale == 0) this else {
      val newPrecision = DecimalType.bounded(precision - scale + 1, 0).precision
      toPrecision(newPrecision, 0, ROUND_FLOOR).getOrElse(
        throw new AnalysisException(s"Overflow when setting precision to $newPrecision"))
    }